### PR TITLE
Add `push.enabled` option to disable push notification calculation

### PR DESCRIPTION
**Pull request by Half-Shot**
Fixes https://github.com/matrix-org/synapse/issues/9524

This should short circuit calls to generate push notifications, for homeservers which don't make practical use of them. This is intended for the likes of things like t2bot.io and libera.chat.

### CODE REVIEW

**Comment by Half-Shot** [comment]
Yish, this is more of a brute force "I don't want notifs to be calculated for any user ever" and that MSC is more opting out some users from notification calculations. 

I think the only practical difference is:

- This PR would save some DB calls to fetch the empty rulesets
- This PR would apply to all users regardless of their settings.

I'm personally still a bit partial to this solution existing in tandem with the MSC, as it still might be useful for debugging / performance purposes (we've disable notification calc via emergency patching in EMS before to debug performance issues and reduce strain in exceptional circumstances).

I will concede that there is overlap though.

File: docs/usage/configuration/config_documentation.md [hunk 1]
```diff
@@ -3355,6 +3355,10 @@ Configuration settings related to push notifications
 This setting defines options for push notifications.
 
 This option has a number of sub-options. They are as follows:
+* `enable_push`: Enables or disables push notification calculation. Note, disabling this will also
+   stop unread counts being calculated for rooms. This mode of operation is intended
+   for homeservers which may only have bots or appservice users connected, or are otherwise
+   not interested in push/unread counters. This is enabled by default.
```
Comment (MatMaul) [inline docs/usage/configuration/config_documentation.md:3361]:
```suggestion
   not interested in push/unread counters. Push is enabled by default.
```

The wording doesn't really tell if you are talking about push itself or the option `enable_push`.

**Review by MatMaul — CHANGES_REQUESTED** [review]

File: docs/usage/configuration/config_documentation.md (hunk 1 above)
Comment (MatMaul) [reply docs/usage/configuration/config_documentation.md:3361]: Sorry my comment is non sense since this is called `enable` and not `disable`...

**Review by MatMaul — APPROVED** [review]

**Comment by DMRobertson** [comment]
Fixup in https://github.com/matrix-org/synapse/pull/14619

**Pull request merged**